C2LC-252: Specify the home position for the character

chosww opened the pull request.

the-t-in-rtf commented:
See [C2LC-252](https://issues.fluidproject.org/browse/C2LC-252) for details.

the-t-in-rtf commented:
As usual I started by trying out the preview site.  A few observations:

First, if you move the character to a corner, point the character diagonally at the corner, and move forward, a very short path is drawn, basically a dot.  I'm not sure this is desirable.  We should discuss it with Sepi and others.

The arrows and turns work well, and the edge behaviour works well.  The only concern I have with the controls to manually input the value of the row or column.  A change in value doesn't take effect until you leave the input.  I would expect a carriage return to immediately attempt to update the value, as you would with a search form or other text input.

This lack of support for carriage returns makes it more confusing when an invalid value is entered.  I remember the discussions about changing the value back to the last valid value, but taken together, it's possible to enter an invalid row or column and not realise that it's invalid until you tab away.

I think adding an update on carriage return would be enough, as then the valid would "snap" back to the last value as a kind of immediate feedback.

the-t-in-rtf commented on src/Character.js:
```diff
@@ -47,10 +55,25 @@ export default class Character extends React.Component<CharacterProps, {}> {
+
+    componentDidMount() {
+        if (this.characterRef.current.scrollIntoView) {
+            this.characterRef.current.scrollIntoView({ behavior: 'auto', block: 'nearest', inline: 'nearest' });
```
If we never use different options, we should find a way to reuse them cleanly, so that they can't get out of sync if we need to change anything.  Defining a constant or creating a small helper function would be fine.

chosww replied:
This change was introduced in 286, so we should resolve any issue to this in 286; however, in Character component, the scrollIntoView property object is used in two places, and they have different scopes, so we can't have a variable for the object, unless we specify that in a class property.

the-t-in-rtf replied:
Thanks for the explanation.  Given that the Character only uses the function in a single way, a class property would seem to make sense.

the-t-in-rtf commented on src/Character.test.js:
```diff
@@ -49,3 +49,30 @@ describe('Right character should render based on theme props', () => {
+    // When the component mounts, call scrollIntoView
+    expect(mockScrollIntoView.mock.calls.length).toBe(1);
+    expect(mockScrollIntoView.mock.calls[0][0]).toStrictEqual({
+        behavior: 'auto',
```
This should also be reused, here it makes sense to define a constant for the expected value.

the-t-in-rtf commented on src/CharacterState.js:
```diff
@@ -92,28 +95,112 @@ export default class CharacterState {
-            y2: this.yPos + yOffset
+            x1: startingX1,
+            y1: startingY1,
+            x2: newXPos,
```
Side note, I like this much better than performing the math inline.

the-t-in-rtf commented on src/CharacterState.js:
```diff
@@ -122,7 +209,111 @@ export default class CharacterState {
+    }
+
+
+    getRowLabel(): string {
```
I'm a little uneasy having the type `string` for the return values here.  It seems like we could protect ourselves better by enumerating the valid values in `types.js`.

chosww replied:
Currently, expected row label values are from 1 to 16, and this may change between any values from 1 to 26, based on our decision on how many rows we want in the scene. Specifying all those values as a type might be undesirable.

the-t-in-rtf replied:
As long as it's clearly identified as providing a string to be used as a label, it's fine.  We can talk about adding more checks if we need them to help avoid actual issues.

the-t-in-rtf commented on src/ProgramBlockEditor.js:
```diff
@@ -35,6 +43,9 @@ type ProgramBlockEditorProps = {
     addNodeExpandedMode: boolean,
     theme: string,
+    onChangeCharacterPosition: (direction: ?string) => void,
+    onChangeCharacterXPosition: (columnLabel: string) => void,
```
The type should also be used here.

the-t-in-rtf commented on src/ProgramBlockEditor.js:
```diff
@@ -523,11 +589,17 @@ class ProgramBlockEditor extends React.Component<ProgramBlockEditorProps, Progra
         if (this.props.theme === 'space') {
-            return <SpaceShipIcon className='ProgramBlockEditor__character-column-character' />
+            return <SpaceShipIcon
+                transform={`rotate(${this.props.characterState.getDirectionDegrees() - 90} 0 0)`}
```
In my own work in this area, I thought I had seen a variation where the transformation was defined as a variable and reused.  That seems more appropriate.

chosww replied:
Good idea, I used a const for the transform.

the-t-in-rtf commented on src/ProgramBlockEditor.js:
```diff
@@ -570,14 +650,92 @@ class ProgramBlockEditor extends React.Component<ProgramBlockEditorProps, Progra
+                            aria-disabled={this.props.editingDisabled}
+                            type='text'
+                            value={this.state.characterColumnLabel}
+                            onChange={!this.props.editingDisabled ? this.handleChangeXPos : undefined}
```
Given that the change doesn't take effect until you blur the control, I'm not sure that `onChange` is adding anything.  In fact you might be attempting to update the position twice.

As discussed elsewhere, we should have support for hitting enter as well.  I think `onBlur` and a small `onKeyDown` handler would be good, unless you have a concrete reason for using `onChange` related to a particular browser.

chosww replied:
onChange handler is responsible for updating values represented on the text box UI, but that doesn't actually change values of CharacterState. We don't want to apply changes as users change text box values, because that will send potentially invalid values for xPos and yPos in CharacterState. OnBlur will communicate with CharacterState state in App and update xPos and yPos in CharacterState after input validation.

the-t-in-rtf commented:
I left a few comments on the code, but found a fairly big blocker afterwards.  The character is not visible at all on the scene in Firefox.  The path is visible, and based on that the position does appear to be correct, but the character is never visible.

the-t-in-rtf commented on src/svg/TurnPositionLeft.svg:
```diff
@@ -0,0 +1,16 @@
+<path fill-rule="evenodd" clip-rule="evenodd" d="M4.10961 1C4.5365 1 4.94591 1.16462 5.24777 1.45765C5.54963 1.75067 5.71921 2.1481 5.71921 2.5625V5.84531C7.05847 4.51917 8.72343 3.54579 10.5573 3.01685C12.3911 2.48791 14.3335 2.42081 16.2014 2.82188C18.0693 3.22296 19.8014 4.079 21.2343 5.30937C22.6673 6.53975 23.7541 8.10397 24.3922 9.85469C24.4697 10.0496 24.506 10.2576 24.4992 10.4665C24.4923 ...
+</g>
+<defs>
+<filter id="filter0_d" x="0.5" y="0" width="26" height="14.9793" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
```
Given our recent experience with the play and stop buttons, we should remove any filters we can.  Do these add anything?

chosww replied:
<img width="142" alt="Screen Shot 2021-03-17 at 1 15 22 PM" src="https://user-images.githubusercontent.com/22668696/111509416-17edd600-8723-11eb-9523-6ff75eb9e849.png">

the-t-in-rtf replied:
Makes sense, then.

the-t-in-rtf commented on src/svg/TurnPositionRight.svg:
```diff
@@ -0,0 +1,16 @@
+<path fill-rule="evenodd" clip-rule="evenodd" d="M22.8904 1C22.4635 1 22.0541 1.16493 21.7522 1.45851C21.4504 1.75209 21.2808 2.15026 21.2808 2.56545V5.85445C19.9415 4.52581 18.2766 3.55059 16.4427 3.02065C14.6089 2.49071 12.6665 2.42349 10.7986 2.82532C8.93065 3.22715 7.19863 4.0848 5.76566 5.3175C4.33269 6.5502 3.24594 8.11736 2.60776 9.87139C2.53034 10.0667 2.49397 10.2751 2.50081 10.4843C2.50 ...
+</g>
+<defs>
+<filter id="filter0_d" x="0.5" y="0" width="26" height="15" filterUnits="userSpaceOnUse" color-interpolation-filters="sRGB">
```
Same question here, is this filter necessary?

chosww commented:
Thanks for reviewing the PR @the-t-in-rtf, I have made updates to address some issues you have found.

- Fixed an issue where the character draws a dot when trying to move outside of the scene from the edge.
- Fixed an issue where the character draws unexpected path when trying to move outside of the scene.
- Implemented new keyDown property on the coordinate text boxes to update the character position on enter key.
- Updated transform of the character in the character column to use const transform variable.

Some changes you suggested regard to the Character component should be resolved in C2LC-286, which Simon has been reviewing. I also want to see what Simon thinks about defining event specific object as a class property, before making the changes.
